### PR TITLE
Revamp #480 restructure

### PR DESCRIPTION
**Pull request by MichaelStritt**
### Linked issue

Check out #480

### How to test

Pray and hope that I broke nothing 👍 

### Comments

I think this should be enough for now. We can do further restructuring later on 👍 

**Edit 28.05.2021**: I just ran the test dataset on low quality and there everything seems to work. Could be really painful to check all the individual commits, but the edits are pretty straight forward. The only real code changes are here f6f5253 and here bef5bf4 .

### CODE REVIEW

**Review by jan-petr — COMMENTED** [review]
Looks like a lot of changes. Great work, though my worries of course dampen my enthusiasm :)

Probably pointless to ask if you always searched for all occurrences to make sure this was replaced everywhere.

If this was done, then I only see a single possibility that you have missed some parameters. And that is if something was changed between the branching and future merging. In these cases, you probably want to scrutinize these commits and make sure that all these are added also to all things modified in between. Note that some can be identified by conflicts, but new code additions won't. That might be some amount of work - what do you think, how much commits were there inbetween?

**Comment by MichaelStritt** [comment]
> 
> 
> Looks like a lot of changes. Great work, though my worries of course dampen my enthusiasm :)
> 
> Probably pointless to ask if you always searched for all occurrences to make sure this was replaced everywhere.
> 
> If this was done, then I only see a single possibility that you have missed some parameters. And that is if something was changed between the branching and future merging. In these cases, you probably want to scrutinize these commits and make sure that all these are added also to all things modified in between. Note that some can be identified by conflicts, but new code additions won't. That might be some amount of work - what do you think, how much commits were there inbetween?

The branch was created four days ago and all of it was implemented within the last four days as well. Most of the changes that happened within the last four days within develop were import related, so I don't see a big issue there. It is a really tricky issue though. Letting it grow older makes things worse though :D

**Comment by jan-petr** [comment]
> turned on the indexing on windows to search within files. This way it was pretty easy to search for all occurrences. Searching for `x.VARIABLE` is pretty easy, but

Agree with all. Yes - all the latest commits/edits were import related, so hopefully nothing really conflicting. But still - do we want to check to be sure? Or we quickly merge? Has Henk seen it? Better ask now and not receive questions later ;)

**Comment by MichaelStritt** [comment]
> 
> 
> > turned on the indexing on windows to search within files. This way it was pretty easy to search for all occurrences. Searching for `x.VARIABLE` is pretty easy, but
> 
> Agree with all. Yes - all the latest commits/edits were import related, so hopefully nothing really conflicting. But still - do we want to check to be sure? Or we quickly merge? Has Henk seen it? Better ask now and not receive questions later ;)

Already checked it. The only commits in between were for #600, #603 & #612. The files that were changes are `xASL_bids_BIDS2Legacy.m`, `xASL_imp_BIDS2Legacy.m ` & `ExploreASL_Initialize.m`. The only conflict should be in `ExploreASL_Initialize.m`. There were no changes in the same lines though.
Sure, we can ask @HenkMutsaerts. I'm just not sure if he has time today and it would be really helpful to merge this quickly, because merging other stuff in between is quite problematic.

**Review by HenkMutsaerts — CHANGES_REQUESTED** [review]
* Is `x.dir` the dataset-related folders (currently in `x.P`)? Or environment/ExploreASL-related folders (currently in `x.D`)?
* Should `x.tools` be `external`?
* What is the difference between `x.opts` and `x.settings`? Input arguments travel through the pipeline right
Also:
* WBmask -> belongs to statistics
* Skull -> belongs to vis
* What is the use for `x.utils`?
* Do you want settings to be separated per module? If yes, then `BILAT_FILTER` is ASL module
* `Pediatric_Template` is an atlas setting?

**Comment by MichaelStritt** [comment]
>     Is x.dir the dataset-related folders (currently in x.P)? Or environment/ExploreASL-related folders (currently in x.D)?

I used it for global directories, like the dataset root, the studyPar, sourceStructure etc. and I thought that those rather complex directories related to mutex etc. are better there than in x.P or x.D, especially since those ones get pretty overfilled and unstructured in the end.

>     Should x.tools be external?

Personal choice, like all of the other fields as well. In the document within the revamp drive and within the ABBA meeting about that we agreed on tools for toolboxes and for me SPM is a toolbox.

>     What is the difference between x.opts and x.settings? Input arguments travel through the pipeline right

x.opts are only the input options of exploreasl master and it's close derivatives right now. x.settings are general pipeline settings that were in exploreasl before and that were in the main level of the x struct before. I think it's a lot cleaner the way it is now and we can easily improve minor things later on.

>     WBmask -> belongs to statistics
>     Skull -> belongs to vis

For me both of those belong to something like `x.masks` or `x.utilities`, but sure, if you insist on that. I guess with statistics you mean x.S ?

>     Do you want settings to be separated per module? If yes, then BILAT_FILTER is ASL module

No, we're not there yet. We need to do this step by step and make sure that the pipeline stays stable. After that we can think about further restructuring.

>     Pediatric_Template is an atlas setting?

Is it? I only know that it's used in the structural module and defined in the data dependent settings.

**Comment by HenkMutsaerts** [comment]
> > Is x.dir the dataset-related folders (currently in x.P)? Or environment/ExploreASL-related folders (currently in x.D)?
> 
> I used it for global directories, like the dataset root, the studyPar, sourceStructure etc. and I thought that those rather complex directories related to mutex etc. are better there than in x.P or x.D, especially since those ones get pretty overfilled and unstructured in the end.

That sounds like `x.D`. Then I would replace all `x.D` with `x.dir`.

> > Should x.tools be external?
> 
> Personal choice, like all of the other fields as well. In the document within the revamp drive and within the ABBA meeting about that we agreed on tools for toolboxes and for me SPM is a toolbox.
 
* True, though we already have a folder called `External`. Always helpful to use the same terms, not using a table to go from "folder terminology" to "internal memory terminology" ;)

> > What is the difference between x.opts and x.settings? Input arguments travel through the pipeline right
> 
> x.opts are only the input options of exploreasl master and it's close derivatives right now. x.settings are general pipeline settings that were in exploreasl before and that were in the main level of the x struct before. I think it's a lot cleaner the way it is now and we can easily improve minor things later on.

* Sure but can becoming confusing because the original settings can also contain input arguments, like which module to run etc?

> > WBmask -> belongs to statistics
> > Skull -> belongs to vis
> 
> For me both of those belong to something like `x.masks` or `x.utilities`, but sure, if you insist on that. I guess with statistics you mean x.S ?

* The masks are used for statistics and visualization. But a good idea to create a `x.masks` or `x.stats.masks`! As long as it stays readable :)
* I would replace `x.S` by `x.stats` for consistency
* What is utilities? This sounds a bit too general. How does this differ from settings?
* skull is a mask as well, used for visualization, so could also go to `x.stats.masks`


> > Do you want settings to be separated per module? If yes, then BILAT_FILTER is ASL module
> 
> No, we're not there yet. We need to do this step by step and make sure that the pipeline stays stable. After that we can think about further restructuring.
> 
> > Pediatric_Template is an atlas setting?
> 
> Is it? I only know that it's used in the structural module and defined in the data dependent settings.

* True, each template is an atlas but it might be used for a different goal indeed. Here the pediatric atlas/template is used for improving registration of young kids, whereas later in the population module these can be used for improved definition of ROIs

**Comment by jan-petr** [comment]
@MichaelStritt I will let you resolve all the errors by Henk. At the same time, I will run some tests and check the code as well for mistakes. But I will report them in comments to avoid conflicting edits...

**Comment by MichaelStritt** [comment]
That's it for me. I don't think it makes sense to change even more now. There's a [follow up issue](https://github.com/ExploreASL/ExploreASL/issues/614) to improve this later on.

**Review by jan-petr — CHANGES_REQUESTED** [review]
- `xASL_adm_CleanUpBeforeRerun.m` Line 369-375: `SPMdir` and `SPMpath` are in a subfield.

**Review by jan-petr — CHANGES_REQUESTED** [review]
Functions/xASL_init_DefineDataDependentSettings.m:39:if ~isfield(x,'Pediatric_Template')

**Review by jan-petr — CHANGES_REQUESTED** [review]
- Functions/xASL_adm_CompareDataSets.m:40:    if  isfield(x,'bReproTesting')
- Functions/xASL_init_DefineDataDependentSettings.m:25:if ~isfield(x,'bReproTesting')
- Functions/xASL_init_DefineDataDependentSettings.m:30:if isfield(x, 'bReproTesting')
- Functions/xASL_Iteration.m:35:          if ~isfield(x,'dryRun')
- Modules/SubModule_ASL/xASL_wrp_ProcessM0.m:83:if ~isfield(x,'M0_conventionalProcessing')
- Modules/xASL_module_Structural.m:347:if ~isfield(x,'SegmentSPM12')
- ExploreASL_ProcessMaster.m:46:        if isfield(x,'SegmentSPM12')
- Functions/xASL_init_DefineDataDependentSettings.m:72:if ~isfield(x, 'SegmentSPM12')
- Functions/xASL_init_DefineDataDependentSettings.m:75:    x.SegmentSPM12 = x.Segment_SPM12;

**Comment by jan-petr** [comment]
> That's it for me. I don't think it makes sense to change even more now. There's a [follow up issue](https://github.com/ExploreASL/ExploreASL/issues/614) to improve this later on.

Agree - the others need a more discussion. And this is all internal, so can be done in two steps. However - please see all the errors I found. They look to me like forgotten renames.

It is mostly isfield(x,'field') instead of isfield(x.subgroup,'field'). And note that in some cases, you might need to add isfield(x,'subgroup'). I keep on searching...

**Comment by MichaelStritt** [comment]
Damn, I thought I found all the isfield ones. They are a bit tricky with the search function. I can fix that 👍

**Comment by MichaelStritt** [comment]
> 
> 
>     * `xASL_adm_CleanUpBeforeRerun.m` Line 369-375: `SPMdir` and `SPMpath` are in a subfield.

Removed them from the list. They are in x.D now ae1eb76. It didn't crash because if this though, right?

**Review by jan-petr — CHANGES_REQUESTED** [review]
- Functions/xASL_Iteration.m:46:          if ~isfield(x,'stopAfterErrors') 
- DataParTemplate.json:5:                 "exclusion": "",
- Functions/xASL_init_DefineStudyData.m:93:if ~isfield(x,'exclusion')
- Functions/xASL_init_DefineStudyData.m:148:if     ~isfield(x,'exclusion')
-

**Comment by jan-petr** [comment]
> Damn, I thought I found all the isfield ones. They are a bit tricky with the search function. I can fix that

I have a better search :)

**Comment by MichaelStritt** [comment]
> > Functions/xASL_init_DefineDataDependentSettings.m:39:if ~isfield(x,'Pediatric_Template')

Hope it's better with 3300654 now 👍

**Comment by HenkMutsaerts** [comment]
Jan should do a find & replace course ;)

**Review by HenkMutsaerts — APPROVED** [review]

**Review by jan-petr — CHANGES_REQUESTED** [review]
- Functions/xASL_adm_CleanUpBeforeRerun.m:370:                        'SPMDIR', 'SPMpath', 'STRUCT_TEMPLATE_IM', 'LOCKDIR', 'SUBJECTDIR',...

File: Functions/xASL_adm_CleanUpBeforeRerun.m
Comment (jan-petr) [inline Functions/xASL_adm_CleanUpBeforeRerun.m:370]: Good that you removed the SPMDIR and SPMpath there, but should you add removal of these fields from the subfields?

**Review by jan-petr — CHANGES_REQUESTED** [review]

**Comment by jan-petr** [comment]
> > > Functions/xASL_init_DefineDataDependentSettings.m:39:if ~isfield(x,'Pediatric_Template')
> 
> Hope it's better with [3300654](https://github.com/ExploreASL/ExploreASL/commit/3300654106e9591355121754afd438dad19475c0) now

Yes - very good :)

File: Functions/xASL_adm_CleanUpBeforeRerun.m
Comment (MichaelStritt) [reply Functions/xASL_adm_CleanUpBeforeRerun.m:370]: I don't think so. x.D is removed there and if SPMDIR and SPMpath are in D, then they will be removed as well?

**Comment by MichaelStritt** [comment]
> 
> 
>     * Functions/xASL_adm_CompareDataSets.m:40:    if  isfield(x,'bReproTesting')
> 
>     * Functions/xASL_init_DefineDataDependentSettings.m:25:if ~isfield(x,'bReproTesting')
> 
>     * Functions/xASL_init_DefineDataDependentSettings.m:30:if isfield(x, 'bReproTesting')
> 
>     * Functions/xASL_Iteration.m:35:          if ~isfield(x,'dryRun')
> 
>     * Modules/SubModule_ASL/xASL_wrp_ProcessM0.m:83:if ~isfield(x,'M0_conventionalProcessing')
> 
>     * Modules/xASL_module_Structural.m:347:if ~isfield(x,'SegmentSPM12')
> 
>     * ExploreASL_ProcessMaster.m:46:        if isfield(x,'SegmentSPM12')
> 
>     * Functions/xASL_init_DefineDataDependentSettings.m:72:if ~isfield(x, 'SegmentSPM12')
> 
>     * Functions/xASL_init_DefineDataDependentSettings.m:75:    x.SegmentSPM12 = x.Segment_SPM12;

Fixed in aa60a31 👍

**Comment by MichaelStritt** [comment]
>     * Functions/xASL_Iteration.m:46:          if ~isfield(x,'stopAfterErrors')
> 
>     * DataParTemplate.json:5:                 "exclusion": "",
> 
>     * Functions/xASL_init_DefineStudyData.m:93:if ~isfield(x,'exclusion')
> 
>     * Functions/xASL_init_DefineStudyData.m:148:if     ~isfield(x,'exclusion')

Fixed in 6e16452 👍

**Comment by MichaelStritt** [comment]
> 
> 
>     * Functions/xASL_adm_CleanUpBeforeRerun.m:370:                        'SPMDIR', 'SPMpath', 'STRUCT_TEMPLATE_IM', 'LOCKDIR', 'SUBJECTDIR',...

SPMDIR and SPMpath were already removed (moved to x.D). SUBJECTDIR was moved in 9a831b5 👍

File: Functions/xASL_Iteration.m
Comment (jan-petr) [inline Functions/xASL_Iteration.m:35]: case - dryrun

File: Functions/xASL_Iteration.m
Comment (jan-petr) [inline Functions/xASL_Iteration.m:38]: case - dryrun

File: Functions/xASL_adm_CleanUpBeforeRerun.m
Comment (jan-petr) [reply Functions/xASL_adm_CleanUpBeforeRerun.m:370]: Of course - you are right :) I didn't realize that...

**Review by jan-petr — CHANGES_REQUESTED** [review]
Almost there - one last tiny bit.

And a tricky question - what happens if somebody processes half of his dataset, then update ExploreASL, and then continues with the rest of the processing. Some parameters will be saved and reloaded from the x.mat and these will now have incorrect names. In the good case, it will crash, in the worst case, it will reinitialize them with default values thus ignoring previous settings. Bottom line - trouble.

My proposition - we need to identify all places where we load x.mat from disk (or datapartemplate because we have the x.dataset.exclusion) and we have to run a function that does all the renaming).
1. Checks if a value with old name is there.
2. Rename to old name and report that this is deprecated and will be removed in future.
3. If the old and new are both present and diverging values - then we report another warning.

File: Functions/xASL_Iteration.m
Comment (MichaelStritt) [reply Functions/xASL_Iteration.m]: Should be fixed now 7e172fb 👍

File: Functions/xASL_Iteration.m
Comment (MichaelStritt) [reply Functions/xASL_Iteration.m]: Should be fixed now 7e172fb 👍

**Comment by MichaelStritt** [comment]
> Almost there - one last tiny bit.
> 
> And a tricky question - what happens if somebody processes half of his dataset, then update ExploreASL, and then continues with the rest of the processing. Some parameters will be saved and reloaded from the x.mat and these will now have incorrect names. In the good case, it will crash, in the worst case, it will reinitialize them with default values thus ignoring previous settings. Bottom line - trouble.
> 
> My proposition - we need to identify all places where we load x.mat from disk (or datapartemplate because we have the x.dataset.exclusion) and we have to run a function that does all the renaming).
> 
> 1. Checks if a value with old name is there.
> 
> 2. Rename to old name and report that this is deprecated and will be removed in future.
> 
> 3. If the old and new are both present and diverging values - then we report another warning.

I just checked the `DataParTemplate.m`. 

I moved `x.SegmentSPM12` to `x.settings.SegmentSPM12`, `x.M0_conventionalProcessing` to `x.settings.M0_conventionalProcessing` and `x.exclusion` to `x.dataset.exclusion`.

@jan-petr: are there other parameters that require a warning?

Full list of changed parameters:
- x.SpaghettiDir
- x.HistogramDir
- x.StatsMaps
- x.SPMDIR
- x.skull
- x.WBmask
- x.SPMpath
- x.SPMVERSION
- x.BILAT_FILTER
- x.Pediatric_Template
- x.bReproTesting
- x.bOverwrite
- x.dryRun
- x.M0_conventionalProcessing
- x.SegmentSPM12
- x.bAutoACPC
- x.bLesionFilling
- x.stopAfterErrors
- x.TotalInclusionList
- x.nExcluded
- x.nTimePointsTotal 
- x.nTimePointTotalSubjects
- x.nTimePointSubjects
- x.TimePointExcluded
- x.ExcludedSubjects 
- x.exclusion
- x.nTimePoints
- x.TotalSubjects 
- x.RERUN
- x.MUTEXID
- x.LockDir
- x.SUBJECTDIR

**Comment by jan-petr** [comment]
I have made all the updates. Found 2 bugs. And tested this with data run with an old version. All seems to be OK and running. @MichaelStritt  - you want to quickly check and test before merging?

**Review by jan-petr — APPROVED** [review]
I approve. But someone should approve my latest changes :)

**Comment by jan-petr** [comment]
The conflict seems to be a simple difference in number of spaces...

**Comment by MichaelStritt** [comment]
> 
> 
> I approve. But someone should approve my latest changes :)

Checked the files. Nice fix in `xASL_adm_LoadX.m ` 👍

**Pull request merged**
